packages/affiliates/libs/affiliates/dappmining_v2.js

```diff
@@ -0,0 +1,278 @@
+
+    return (
+      highland(stream)
+        // trace transactions may contain the contract creation event, we cant decode this so filter out
+        // fyi non contract creation "trace_type" are called "call"
+        .filter(tx => tx.trace_type !== "create" && tx.error == null)
+        .map(tx => {
+          return decodeTx(tx, {
+            blockNumber: tx.block_number,
+            blockTimestamp: moment(tx.block_timestamp.value).valueOf(),
+            ...tx
+          });
+        })
+        // this filters only emp contract "create" calls. We could technically look at more, but these
+        // would be the only "attributed" calls as per the spec
+        .filter(tx => tx.name == "create")
+    );
+  }
+
+  // returns the proportion of rewards given to each developer for a specific point in time
+  function calculateBlockReward(
+    { attributions, balances, whitelist },
+    // return all tags with a percent value for rewards
+    rewards = {}
+  ) {
+    // for every user with a token balance
+    balances.keys().forEach(userid => {
+      // get users balance
+      const balance = balances.get(userid);
+      const attributionPercents = attributions.getAttributionPercents(userid, balance);
+      // go through all whitelisted tags
+      whitelist.forEach(tag => {
+        if (rewards[tag] == null) rewards[tag] = toBN("0");
+        // get that tags attribution percent for that user relative to all other affiliates
+        const attributionPercent = attributionPercents[tag] || "0";
+        // multiply the balance and the attribution percent to get final attribution weight
+        // and add it to the rest of the weights. This will need to be divided by total balances eventually.
+        rewards[tag] = rewards[tag].add(toBN(balance).mul(toBN(attributionPercent)));
+      });
+    });
+    // normalize balances based on total in all balances. This is done last instead of in the above to
+    // reduce precision loss which seems to happen the more divisions there are.
+    const total = toBN(balances.getTotal());
+    return Object.entries(rewards).reduce((result, [key, value]) => {
+      result[key] = value.div(total).toString();
+      return result;
+    }, {});
+  }
+
+  // integer percent calculation where percent is passed in as wei
+  function calculatePercent(amount, percent) {
+    return fromWei(toBN(amount.toString()).mul(toBN(percent.toString())));
+  }
+
+  // calculates attribution weights for multiple blocks by simply taking last snapshot state and multiplying by
+  // the number of blocks elapsed.
+  function sumAttributions(attributions, blocksElapsed = 1, sum = AttributionLookback()) {
+    blocksElapsed = toBN(blocksElapsed);
+    attributions.forEach((userid, affiliate, amount, index) => {
+      let currentSum;
+      try {
+        currentSum = sum.getByIndex(userid, index).amount;
+      } catch (err) {
+        currentSum = "0";
+      }
+      const weighted = toBN(amount)
```

Comment:
I feel like I'm missing something in this calculation. Why do you weight by number of blocks elapsed?

Reply:
i made an observation about the math for dapp mining ( and could apply to dev mining), that allowed the calculation to be optimized. 

the first way we can do this calculation is that we must know the state of the system at every block, this includes every user balance, all attributions and the per block reward. We then accumulate the rewards at every block until we get to the last block. This is hard to do without iterating and snapshotting the state at every block once, then iterating again to attribute rewards. This is exactly what dev mining does. 

In dapp mining, the math works out the same if we basically sum up the state of balances and attributions independently and apply the rewards calculation at the very end. You can imagine if someone's balance was 100 at block 1 though 10, this basically says their balance weight for 10 blocks is 100 * 10. Its the same thing for attributions, these will also be accumulated every block any time there is a state change.  

hopefully this makes sense!

Reply:
Yep, it does!

Reply:
this is pretty awesome.